linalg/solvers.hpp

```diff
@@ -65,24 +65,111 @@ class IterativeSolverMonitor
+      /// Detailed information about each iteration will be reported
+      bool iterations = false;
+      /** @brief A summary of the solver process will be reported after the last
+          iteration */
+      bool summary = false;
+      /// @brief Information about the first and last iteration will be printed
+      bool first_and_last = false;
+
+      /// Initializes the print level to suppress
+      PrintLevel() = default;
+
+      /** @name Builder
+         These methods are utilized to construct PrintLevel objects through a
+         builder approach by chaining the function calls in this group. */
+      ///@{
+      PrintLevel &None() { *this = PrintLevel(); return *this; }
+      PrintLevel &Warnings() { warnings=true; return *this; }
+      PrintLevel &Errors() { errors=true; return *this; }
+      PrintLevel &Iterations() { iterations=true; return *this; }
+      PrintLevel &FirstAndLast() { first_and_last=true; return *this; }
+      PrintLevel &Summary() { summary=true; return *this; }
+      PrintLevel &All() { return Warnings().Errors().Iterations().FirstAndLast().Summary(); }
+      ///@}
+   };
+
 #ifdef MFEM_USE_MPI
 private:
    int dot_prod_type; // 0 - local, 1 - global over 'comm'
-   MPI_Comm comm;
+   MPI_Comm comm = MPI_COMM_NULL;
 #endif
 
 protected:
    const Operator *oper;
    Solver *prec;
    IterativeSolverMonitor *monitor = nullptr;
 
-   int max_iter, print_level;
-   double rel_tol, abs_tol;
+   /** @name Reporting
+      These options control the internal reporting behavior into ::mfem::out
+      and ::mfem::err of the iterative solvers. */
+   ///@{
+
+   /** @brief Legacy print level definition, which is left for compatibility
+       with custom iterative solvers. #print_options should be used instead. */
+   MFEM_DEPRECATED int print_level = -1;
+
+   /** @brief Output behavior for the iterative solver.
+
+       This primarily controls the output behavior of the iterative solvers
+       provided by this library. This member must be synchronized with
+       #print_level to ensure compatibility with custom iterative solvers. */
+   PrintLevel print_options;
+
+   /// Convert a legacy print level integer to a PrintLevel object
+   PrintLevel FromLegacyPrintLevel(int);
+
+   /// @brief Use some heuristics to guess a legacy print level corresponding to
+   /// the given @a PrintLevel.
+   static int GuessLegacyPrintLevel(PrintLevel);
+   ///@}
+
+   /** @name Convergence
+       @brief Termination criteria for the iterative solvers.
+
+       @a X denotes the space in which the norm is measured, whose choice
+       depends on the specific iterative solver. */
+   ///@{
+
+   /// Limit for the number of iterations the solver is allowed to do
+   int max_iter;
 
-   // stats
-   mutable int final_iter, converged;
+   /// Convergence criterion: \f$ ||r||_X \leq rel_{tol}||r_0||_X \f$
+   double rel_tol;
+
+   /// Convergence criterion: \f$ ||r||_X \leq abs_{tol} \f$
+   double abs_tol;
```

Comment:
These comments are not quite accurate: there is one convergence criterion that uses two tolerance values (I'd call them `tol_{rel}` and `tol_{abs}` in the Latex expression, instead of `rel_{tol}` and `abs_{tol}`):
```
||r||_X \leq \max\{ tol_{abs}, tol_{rel} ||r_0||_X \}
```

Reply:
Thanks for the elaboration. Will update the docs later.

Should I also rename the variables to be consistent or should I leave a todo here? I am asking because technically speaking renaming these variables is a breaking change (although easy to fix for users).

Reply:
No need to change the variable names -- I just think when writing a math expression it is more common to use the noun as the name of the "variable" and the adjective as the subscript: tol<sub>rel</sub>. The connection between tol<sub>rel</sub> and `rel_tol` is still clear, I think.

In any case, the notation is a minor thing -- if you prefer to keep it as it is now -- that is fine too.

Reply:
Since I am already at it I prefer to update it to be as precise as possible. :)

Reply:
When updating this I realized another issue: Doxygen swallows the group docs for some reason. Not sure if it is a problem on my machine only or if I messed up somewhere.

Reply:
Doxygen is hard to understand sometimes. I pushed some changes that should fix the disappearing documentation.

Reply:
Looking at the change you made I am not sure how I messed up this. Tried your changes locally and now everything looks fine. Thanks for taking the time to fix it!